File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0206-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # O(n) algorithm with O(1) extra space
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None: return head

        p = None
        q = head
        r = head.next

        while q != None:
            q.next = p
            p = q
            q = r
            if r != None: r = r.next
        return p 

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/04-Contiguous-LinkedAllocation/LeetCode/LeetCode-0206-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /// O(n) algorithm with O(1) extra space
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode p = null;
        ListNode q = head;
        ListNode r = head.next;

        while (q != null){
            q.next = p;
            p = q;
            q = r;
            if (r != null) r = r.next;
        } //end-while

        return p;              
    } // end-reverseList
}